homeassistant/components/google/api.py

```diff
@@ -2,20 +2,169 @@
-from oauth2client.file import Storage
-
-from homeassistant.core import HomeAssistant
+import oauth2client
+from oauth2client.client import (
+    Credentials,
+    DeviceFlowInfo,
+    FlowExchangeError,
+    OAuth2Credentials,
+    OAuth2DeviceCodeError,
+    OAuth2WebServerFlow,
+)
+
+from homeassistant.const import CONF_CLIENT_ID, CONF_CLIENT_SECRET
+from homeassistant.core import CALLBACK_TYPE, HomeAssistant
+from homeassistant.helpers import config_entry_oauth2_flow
+from homeassistant.helpers.event import async_track_time_interval
 from homeassistant.util import dt
 
-_LOGGER = logging.getLogger(__name__)
+from .const import CONF_CALENDAR_ACCESS, DATA_CONFIG, DEVICE_AUTH_IMPL, DOMAIN
 
+_LOGGER = logging.getLogger(__name__)
 
 EVENT_PAGE_SIZE = 100
+EXCHANGE_TIMEOUT_SECONDS = 60
+
+
+class OAuthError(Exception):
+    """OAuth related error."""
+
+
+class TokenNotReady(Exception):
+    """Raised when the token exchange is not yet ready, and should be attempted again."""
+
+
+class DeviceAuth(config_entry_oauth2_flow.LocalOAuth2Implementation):
+    """OAuth implementation for Device Auth."""
+
+    def __init__(self, hass: HomeAssistant, client_id: str, client_secret: str) -> None:
+        """Initialize InstalledAppAuth."""
+        super().__init__(
+            hass,
+            DEVICE_AUTH_IMPL,
+            client_id,
+            client_secret,
+            oauth2client.GOOGLE_AUTH_URI,
+            oauth2client.GOOGLE_TOKEN_URI,
+        )
+
+    async def async_resolve_external_data(self, external_data: Any) -> dict:
+        """Resolve a Google API Credentials object to Home Assistant token."""
+        creds: Credentials = external_data["creds"]
+        return {
+            "access_token": creds.access_token,
+            "refresh_token": creds.refresh_token,
+            "scope": " ".join(creds.scopes),
+            "token_type": "Bearer",
+            "expires_in": creds.token_expiry.timestamp(),
+        }
+
+
+class DeviceFlow:
+    """OAuth2 device flow for exchanging a code for an access token."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        oauth_flow: OAuth2WebServerFlow,
+        device_flow_info: DeviceFlowInfo,
+    ) -> None:
+        """Initialize DeviceFlow."""
+        self._hass = hass
+        self._oauth_flow = oauth_flow
+        self._device_flow_info: DeviceFlowInfo = device_flow_info
+        self._exchange_task_unsub: CALLBACK_TYPE | None = None
+
+    @property
+    def verification_url(self) -> str:
+        """Return the verification url that the user should visit to enter the code."""
+        return self._device_flow_info.verification_url
+
+    @property
+    def user_code(self) -> str:
+        """Return the code that the user should enter at the verification url."""
+        return self._device_flow_info.user_code
+
+    async def start_exchange_task(
+        self, finished_cb: Callable[[Credentials | None], Awaitable[None]]
```

Comment:
Why is the credentials parameter in the callback optional?

Reply:
The callback is invoked with the valid credentials or with None if the task timed out or the credential expired. There is a comment in the pydoc here (since i knew this was subtle), but I would love to have more suggestions on improving readability if you have them.

Reply:
I wanted to also mention that right now the behavior for these cases are currently treated the same, as they were before:
- token is not ready because the user has not typed in the code yet
- you misconfigured the oauth client id
- server error occurred

And in followup PRs i'd like to have more flexible error codes here.  Perhaps using a real `Task` object here in a follow up PR is the way to go in the future so these various cases can be more easily be expressed.  Generally, this whole integration needs some error code handling improvements to move up the quality scale.